graph/run-dynamic.go

```diff
@@ -263,59 +263,69 @@ func ConvergeDynamicThread[EP EPP[E], V VPI[V], E EPI[E], M MVI[M], N any, A Alg
 				// There is a request for synchronously syncing topology (blocking), acknowledge here as we have no more topology events to process.
 				bspSync = false
 				gt.Response <- ACK
-			} else if remitCount == 0 {
-				// Want topology events, but none found -- we should relax.
-				topFail++
-				gt.Status = BACKOFF_TOP
-				utils.BackOff(topFail)
-				if timeStates && tidx == 0 {
-					curr = time.Now()
-					gt.LoopTimes[5] += curr.Sub(prev)
-					prev = curr
-				}
-			} else {
-				topFail = 0
+			} else if remitCount == 0 && !epoch {
+				topFailed = true
 			}
 
 			if strucClosed {
 				doneEvents <- struct{}{}
-			} else if TOPOLOGY_FIRST || blockAlgIfTop {
-				continue
 			}
 		}
 
-		// Process algorithm messages. Check for algorithm termination if needed.
-		gt.Status = APPLY_MSG
-		checkTerm := (strucClosed && remitClosed) || (epoch && remitCount == 0 && count == 0)
-		completed, count = ProcessMessages[V, E, M, N](alg, g, gt, checkTerm)
-		if completed && epoch {
-			//log.Debug().Msg("T[" + utils.F("%02d", tidx) + "] completed epoch")
-			gt.Status = DONE
-			gt.Response <- ACK
-
-			resp := <-gt.Command // BLOCK and wait for resume
-			if resp != RESUME {
-				log.Panic().Msg("Expected to resume after blocked")
+		algFailed := false
+		if strucClosed || !(TOPOLOGY_FIRST || blockAlgIfTop) {
+			// Process algorithm messages. Check for algorithm termination if needed.
+			gt.Status = APPLY_MSG
+			checkTerm := (strucClosed && remitClosed) || (epoch && remitCount == 0 && topCount == 0)
+			completed, algCount = ProcessMessages[V, E, M, N](alg, g, gt, checkTerm)
+			if completed && epoch {
+				//log.Debug().Msg("T[" + utils.F("%02d", tidx) + "] completed epoch")
+				gt.Status = DONE
+				gt.Response <- ACK
+
+				resp := <-gt.Command // BLOCK and wait for resume
+				if resp != RESUME {
+					log.Panic().Msg("Expected to resume after blocked")
+				}
+				epoch = false
+				completed = false
+			} else if algCount == 0 {
+				algFailed = true
 			}
-			epoch = false
-			completed = false
-			g.TerminateVotes[tidx] = 0
 		}
 
-		if count == 0 {
-			algFail++
-			if algFail%10 == 0 {
-				gt.Status = BACKOFF_ALG
-				utils.BackOff(algFail / 10)
+		// Backoff only when there are no topological events and algorithmic event
+		if algFailed && topFailed {
+			if topFailed {
+				gt.Status = BACKOFF_TOP
+				utils.BackOff(topFail)
 				if timeStates && tidx == 0 {
 					curr = time.Now()
-					gt.LoopTimes[6] += curr.Sub(prev)
+					gt.LoopTimes[5] += curr.Sub(prev)
 					prev = curr
 				}
+				topFail++
+			} else {
+				topFail = 0
+			}
+			if algFailed {
+				if algFail%10 == 0 {
+					gt.Status = BACKOFF_ALG
+					utils.BackOff(algFail / 10)
+					if timeStates && tidx == 0 {
+						curr = time.Now()
+						gt.LoopTimes[6] += curr.Sub(prev)
+						prev = curr
+					}
+				}
+				algFail++
+			} else {
+				algFail = 0
 			}
 		} else {
-			algFail = 0
+			topFail, algFail = 0, 0
```

Comment:
Is this necessary?

Reply:
This supersedes the need for setting them individually above. And is less code. :)

Reply:
I don't think this supersedes the above ones :) If only one of top or alg failed, the failure count of the successful one would not be reset to 0.

Reply:
Hm, I think that should be the behaviour (resetting both), as we only enter the branch above if both were failed (algFailed && topFailed)

Probably can simplify the code further in this case.

Reply:
Do you mean the current logic is ok? Maybe we should keep only one fail count

Reply:
Indeed -- as in this is equivalent to just one back-off that occurs here under one condition. But IIRC there is an important reason to back off on topology separately. 

I guess the ideal way to back-off topology is not to delay, but to process algorithm, but this strategy here is not going to delay topology checking... hm, I need to investigate this more...

Reply:
This also has a bad interaction with static graph loading, but the fix is straightforward...

Reply:
Seems okay so long as the other change I suggested is in place.

So definitely worth simplifying this into a single back-off case.

Reply:
Oh I think a similar issue also exists when `strucClosed == true` or `blockTop == true` - backoff will never be triggered again even when there are no algorithmic messages.

Reply:
Hm, I think I see what you're referring to. Yeah, best to clean up the code further and ensure the logic is correct.
My point is that it should backoff if there was no topology and the algorithm loop was skipped.

I suggest trying to clean up the branches further :)

Reply:
Do you remember the purpose of reducing backoffs by a factor of 10 for alg failures? Should I keep this behaviour?

Reply:
I believe because one of the times the algorithm is here is when its nearing termination and few messages are passed across threads; so while we do want some backoff, we want not much as threads should poll for messages pretty often. 
Its different behaviour to the event stream, as new/more events will just get queued, whereas algorithm messages are generated from other algorithm messages.